web_poet/overrides.py

```diff
@@ -0,0 +1,207 @@
+        exclude: Union[str, Iterable[str], None] = None,
+        priority: int = 500,
+        **kwargs,
+    ):
+        """
+        Class decorator that indicates that the decorated Page Object should be
+        used instead of the overridden one for a particular set the URLs.
+
+        Which Page Object is overridden is determined by the ``overrides``
+        parameter.
+
+        Over which URLs the override happens is determined by the ``include``,
+        ``exclude`` and ``priority`` parameters. See the documentation of the
+        `url-matcher <https://url-matcher.readthedocs.io/>`_ package for more
+        information about them.
+
+        Any extra parameters are stored as meta information that can be later used.
+
+        :param include: Defines the URLs that should be handled by the overridden Page Object.
+        :param overrides: The Page Object that should be replaced by the annotated one.
+        :param exclude: Defines URLs over which the override should not happen.
+        :param priority: The resolution priority in case of conflicting annotations.
+        """
+
+        def wrapper(cls):
+            rule = OverrideRule(
+                for_patterns=Patterns(
+                    include=_as_list(include),
+                    exclude=_as_list(exclude),
+                    priority=priority,
+                ),
+                use=cls,
+                instead_of=overrides,
+                meta=kwargs,
+            )
+            # If it was already defined, we don't want to override it
+            if cls not in self.data:
+                self.data[cls] = rule
+            else:
+                warnings.warn(
+                    f"Multiple @handle_urls annotations with the same 'overrides' "
+                    f"are ignored in the same Registry. Ignoring duplicate "
+                    f"annotation on '{include}' for {cls}."
+                )
+
+            return cls
+
+        return wrapper
+
+    def get_overrides(self) -> List[OverrideRule]:
+        """Returns all override rules that were declared using ``@handle_urls``."""
+        return list(self.data.values())
+
+    def get_overrides_from_module(self, module: str) -> List[OverrideRule]:
+        """Returns the override rules that were declared using ``@handle_urls``
+        in a specific module.
+
+        This is useful if you've organized your Page Objects into multiple
+        submodules in your project as you can filter them easily.
+        """
+        rules: Dict[Callable, OverrideRule] = {}
+
+        for mod in walk_modules(module):
+            # Dict ensures that no duplicates are collected and returned.
+            rules.update(self._filter_from_module(mod.__name__))
+
+        return list(rules.values())
+
+    def _filter_from_module(self, module: str) -> Dict[Callable, OverrideRule]:
+        return {
+            cls: rule
+            for cls, rule in self.data.items()
+            if cls.__module__.startswith(module)
```

Comment:
Can string matching lead to issues here? E.g. cls.__module__ is foo.bar, and module is "fo"?

I also wonder if startswith is necessary; doesn't walk_modules import all modules recursively, so that you can always use exact match?

Caveat: I'm not confident I understood this part correctly :)

Reply:
> Can string matching lead to issues here? E.g. cls.__module__ is foo.bar, and module is "fo"?

Great catch! 🙏  It would indeed have some matching issues here. Fixed it on https://github.com/scrapinghub/web-poet/pull/16/commits/c000cbc9d36f1c3022c22d27d67d83e4d86b41c7.

> I also wonder if startswith is necessary; doesn't walk_modules import all modules recursively, so that you can always use exact match?

What we're filtering out here is the Registry's `data` values which contains the mapping of override rules. There might be cases where the registry was applied with `walk_modules` and could now contain `data` values on some other package/module. This filtration would prevent that.